examples/image_viewer.rs

```diff
@@ -0,0 +1,356 @@
+extern crate image;
+
+use std::env;
+use std::sync::{Arc, Mutex, Weak};
+
+use smithay_client_toolkit::{
+    compositor::{CompositorHandler, CompositorState, SurfaceData, SurfaceDataExt},
+    delegate_compositor, delegate_output, delegate_registry, delegate_shm, delegate_xdg_shell,
+    delegate_xdg_window,
+    output::{OutputHandler, OutputState},
+    registry::{ProvidesRegistryState, RegistryState},
+    shell::xdg::{
+        window::{Window, WindowConfigure, WindowHandler, XdgWindowState},
+        XdgShellHandler, XdgShellState,
+    },
+    shm::{
+        pool::slot::{Buffer, SlotPool},
+        ShmHandler, ShmState,
+    },
+};
+use wayland_client::{
+    delegate_dispatch,
+    protocol::{wl_output, wl_shm, wl_surface},
+    Connection, Proxy, QueueHandle,
+};
+
+fn main() {
+    env_logger::init();
+
+    let conn = Connection::connect_to_env().unwrap();
+
+    let display = conn.display();
+
+    let mut event_queue = conn.new_event_queue();
+    let qh = event_queue.handle();
+
+    let registry = display.get_registry(&qh, ()).unwrap();
+
+    let mut state = State {
+        registry_state: RegistryState::new(registry),
+        output_state: OutputState::new(),
+        compositor_state: CompositorState::new(),
+        shm_state: ShmState::new(),
+        xdg_shell_state: XdgShellState::new(),
+        xdg_window_state: XdgWindowState::new(),
+
+        pool: None,
+        windows: 0,
+    };
+
+    event_queue.blocking_dispatch(&mut state).unwrap();
+    event_queue.blocking_dispatch(&mut state).unwrap();
+
+    let mut pool_size = 0;
+
+    for path in env::args_os().skip(1) {
+        let image = match image::open(&path) {
+            Ok(i) => i,
+            Err(e) => {
+                println!("Failed to open image {}.", path.to_string_lossy());
+                println!("Error was: {:?}", e);
+                return;
+            }
+        };
+
+        // We'll need the image in RGBA for drawing it
+        let image = image.to_rgba8();
+
+        pool_size += image.width() * image.height() * 4;
+
+        let viewer = Arc::new_cyclic(|weak| {
```

Comment:
What's the rationale behind storing all the image state in an arc and sharing it with every window vs using the &mut self reference you have in all the callbacks?

Reply:
Avoiding the requirement that the `&mut self` maintain and walk a list of all surfaces when handling events.

Looking at the individual patches instead of the big diff makes this clearer.

Reply:
I could drop the Demonstrate patch; the example might be clearer without it.